zookeeper-server/src/main/java/org/apache/zookeeper/server/ServerCnxnFactory.java

```diff
@@ -272,4 +277,48 @@ protected void configureSaslLogin() throws IOException {
         }
     }
 
+    /** Maximum number of connections allowed in the ZooKeeper system */
+    public int getMaxCnxns() {
+        return maxCnxns;
+    }
+
+    protected void initMaxCnxns() {
+        String mxCxnProp = System.getProperty(ZOOKEEPER_MAX_CONNECTION);
+        maxCnxns = ZOOKEEPER_MAX_CONNECTION_DEFAULT;
+        if (null != mxCxnProp) {
+            LOG.info("maxCnxns configured value is {}.", mxCxnProp);
+            try {
+                maxCnxns = Integer.parseInt(mxCxnProp.trim());
+            } catch (NumberFormatException e) {
+                LOG.warn("maxCnxns configured value '{}' can not be parsed to integer, using default value {}.",
+                        mxCxnProp, ZOOKEEPER_MAX_CONNECTION_DEFAULT);
+                }
+            if (maxCnxns < 0) {
+                maxCnxns = 0;
+                LOG.warn("maxCnxns should be greater than or equal to 0, using default vlaue {}.",
+                        ZOOKEEPER_MAX_CONNECTION_DEFAULT);
+            }
+
+        } else {
+            LOG.warn("maxCnxns is not configured, using default value {}.",
+                        ZOOKEEPER_MAX_CONNECTION_DEFAULT);
+        }
+    }
+
+    /**
+      * Ensure total number of connections are less than the maxCnxns
+      *
+      * @throws IOException
+      *             When number of connections exceeds maxCnxns limit
+      */
+    protected void limitTotalNumberOfCnxns() throws IOException {
+        if (maxCnxns <= 0) {
+            // maxCnxns limit is disabled
+            return;
+        }
+        int cnxns = getNumAliveConnections();
+        if (cnxns >= maxCnxns) {
+            throw new IOException("Too many connections " + cnxns + " - max is " + maxCnxns);
```

Comment:
A graceful shutdown would be nice. See the use of `ServerMetrics.getMetrics().CONNECTION_REJECTED` in the code base. Alternately, could reuse the `ClientCnxnLimitException` exception.

Reply:
Done, thanks